### PR TITLE
doc/rados: edit crush-map.rst (4 of x)

### PR DESCRIPTION
**Pull request by zdover23**
Edit doc/rados/operations/crush-map.rst.

https://tracker.ceph.com/issues/58485

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [x] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: doc/rados/operations/crush-map.rst [hunk 1]
```diff
@@ -891,211 +891,236 @@ Migration impact:
...
+Client versions that support CRUSH_TUNABLES2
 --------------------------------------------
 
- * argonaut series, v0.48.1 or later
- * v0.49 or later
- * Linux kernel version v3.6 or later (for the file system and RBD kernel clients)
-
-Which client versions support CRUSH_TUNABLES2
----------------------------------------------
+ * v0.55 and later, including bobtail series (v0.56.x)
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:897]: including Bobtail (v0.56.x)

File: doc/rados/operations/crush-map.rst [hunk 2]
```diff
@@ -891,211 +891,236 @@ Migration impact:
...
+ * Linux kernel version v3.9 and later (for the file system and RBD kernel clients)
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:898]: CephFS and RBD

File: doc/rados/operations/crush-map.rst [hunk 3]
```diff
@@ -891,211 +891,236 @@ Migration impact:
...
- * Linux kernel version v3.9 or later (for the file system and RBD kernel clients)
-
-Which client versions support CRUSH_TUNABLES3
----------------------------------------------
+Client versions that support CRUSH_TUNABLES3
+--------------------------------------------
 
- * v0.78 (firefly) or later
- * Linux kernel version v3.15 or later (for the file system and RBD kernel clients)
+ * v0.78 (firefly) and later
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:903]: Firefly

File: doc/rados/operations/crush-map.rst [hunk 4]
```diff
@@ -891,211 +891,236 @@ Migration impact:
...
+ * Linux kernel version v3.15 and later (for the file system and RBD kernel clients)
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:904]: CephFS

File: doc/rados/operations/crush-map.rst [hunk 5]
```diff
@@ -891,211 +891,236 @@ Migration impact:
...
 
-Which client versions support CRUSH_V4
---------------------------------------
+Client versions that support CRUSH_V4
+-------------------------------------
 
- * v0.94 (hammer) or later
- * Linux kernel version v4.1 or later (for the file system and RBD kernel clients)
+ * v0.94 (hammer) and later
+ * Linux kernel version v4.1 and later (for the file system and RBD kernel clients)
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:910]: Hammer, CephFS

File: doc/rados/operations/crush-map.rst [hunk 6]
```diff
@@ -891,211 +891,236 @@ Migration impact:
...
 
-Which client versions support CRUSH_TUNABLES5
----------------------------------------------
+Client versions that support CRUSH_TUNABLES5
+--------------------------------------------
 
- * v10.0.2 (jewel) or later
- * Linux kernel version v4.5 or later (for the file system and RBD kernel clients)
+ * v10.0.2 (jewel) and later
+ * Linux kernel version v4.5 and later (for the file system and RBD kernel clients)
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:916]: Jewel, CephFS

File: doc/rados/operations/crush-map.rst [hunk 7]
```diff
@@ -891,211 +891,236 @@ Migration impact:
...
-You can apply a profile to a running cluster with the command:
+When making adjustments to CRUSH tunables, keep the following considerations in
+mind:
+
+ * Adjusting the values of the CRUSH tunables will result in the shift
+   of one or more PGs from one storage node to another. If the Ceph cluster is
+   already storing a great deal of data, be prepared for a certain amount of
+   data movement.
+ * When the ``ceph-osd`` and ``ceph-mon`` daemons get the updated map,
+   they immediately begin rejecting new connections with clients that do
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:976]: s/with/from/

File: doc/rados/operations/crush-map.rst [hunk 8]
```diff
@@ -891,211 +891,236 @@ Migration impact:
...
-   are hardcoded and are generally a mix of optimal and legacy values.
-   These values generally match the ``optimal`` profile of the previous
-   LTS release, or the most recent release for which we generally expect
-   most users to have up-to-date clients for.
-
-You can apply a profile to a running cluster with the command:
+When making adjustments to CRUSH tunables, keep the following considerations in
+mind:
+
+ * Adjusting the values of the CRUSH tunables will result in the shift
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:971]: s/the //

File: doc/rados/operations/crush-map.rst [hunk 9]
```diff
@@ -891,211 +891,236 @@ Migration impact:
...
+   of one or more PGs from one storage node to another. If the Ceph cluster is
+   already storing a great deal of data, be prepared for a certain amount of
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:973]: s/certain/significant/

File: doc/rados/operations/crush-map.rst [hunk 10]
```diff
@@ -891,211 +891,236 @@ Migration impact:
...
+However, sometimes it is clear that an OSD is not well suited to act as the
+lead as compared with other OSDs (for example, if the OSD has a slow drive or a
+slow controller). To prevent performance bottlenecks (especially on read
+operations) and at the same time maximize the utilization of your hardware, you
+can influence the selection of the primary OSD either by adjusting "primary
+affinity" values, or by crafting a CRUSH rule that selects OSDs that are better
+suited to act as the lead rather than other OSDs.
+
+To determine whether tuning Ceph's selection of primary OSDs will improve
+cluster performance, pool type must be taken into account. For replicated
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:1038]: s/type/redundancy strategy/

File: doc/rados/operations/crush-map.rst [hunk 11]
```diff
@@ -891,211 +891,236 @@ Migration impact:
...
+affinity values, let us suppose that a cluster contains a mix of drive sizes:
+for example, suppose it contains some older racks with 1.9 TB SATA SSDs and
+some newer racks with 3.84 TB SATA SSDs. The latter will on average be assigned
+twice the number of PGs and will thus serve twice the number of write and read
+operations -- they will be busier than the former. In such a scenario, you
+might make a rough assignment of primary affinity as inversely proportional to
+OSD size. Such an assignment will not be 100% optimal, but it can readily
+achieve a 15% improvement in overall read throughput by means of a more even
+utilization of SATA interface bandwidth and CPU cycles.
+
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:1062]: FWIW this was based on an actual cluster.

**Review by anthonyeleven — APPROVED** [review]

File: doc/rados/operations/crush-map.rst (hunk 1 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:897]: Accepted.

File: doc/rados/operations/crush-map.rst (hunk 2 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:898]: Accepted.

File: doc/rados/operations/crush-map.rst (hunk 3 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:903]: Accepted.

File: doc/rados/operations/crush-map.rst (hunk 4 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:904]: Accepted.

File: doc/rados/operations/crush-map.rst (hunk 5 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:910]: Accepted.

File: doc/rados/operations/crush-map.rst (hunk 6 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:916]: Accepted.

File: doc/rados/operations/crush-map.rst (hunk 7 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:976]: Accepted. Good suggestion.

File: doc/rados/operations/crush-map.rst (hunk 8 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:971]: Accepted.

File: doc/rados/operations/crush-map.rst (hunk 11 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:1062]: I'll add a note that lets the reader know as much. I think such remarks give the documentation a reassuring human touch that increases confidence in the work.

File: doc/rados/operations/crush-map.rst (hunk 9 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:973]: Accepted.

File: doc/rados/operations/crush-map.rst (hunk 10 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:1038]: Accepted.

**Comment by zdover23** [comment]
https://github.com/ceph/ceph/pull/52098 - Reef backport
https://github.com/ceph/ceph/pull/52099 - Quincy backport

**Pull request merged**